mldsa-native integration

mkannwischer opened the pull request:
This pull request imports [mldsa-native ](https://github.com/pq-code-package/mldsa-native) - a high-speed high-assurance ML-DSA implementation deveoped as a part of the post-quantum code package (PQCP) and maintained by @hanno-becker, @jakemas, and myself. It includes a portable C implementation, a x86_64 implementation, and an AArch64 implementation. This integration replaces the existing pqcrystals implementations.

This pull request also removes the patches that were required for the pqcrystals implementation. For mldsa-native, we require no patches.
Furthermore, I have removed the constant-time exceptions in tests/constant_time/sig/passes/ml_dsa*. For mldsa-native, we don't require exceptions and instead use the excplicit declassifications (`MLD_CT_TESTING_DECLASSIFY`)


**Performance** We have measured the impact of this integration on liboqs performance on various x86_64 CPUs in [#546](https://github.com/pq-code-package/mldsa-native/issues/546). For the portable implementation we see a solid boost in performance of 9-25% despite conservative constant-time hardening (see [#371](https://github.com/pq-code-package/mldsa-native/pull/371), [#392](https://github.com/pq-code-package/mldsa-native/pull/392), [#391](https://github.com/pq-code-package/mldsa-native/pull/391)). For the x86_64-optimized implementation, there is a small performance regression of up to 13% in signing. We primarily account this  regression to our x86_64 implementation so far not including all AVX2 intrinsics from the pqcrystals implementation. The reason for being selective in pulling in native code is that we strive to formally verify the correctness of each native function using HOL-Light later on (this is still work on progress). We will revisit that x86_64 performance gap in [#545](https://github.com/pq-code-package/mldsa-native/issues/545) and hope to close this gap in a follow-up release. Likewise, for AArch64 we are expecting some small performance improvements over time.

**Formal verifcation** All C code in the portable C implementation is proven type-safe and memory-safe using the C Bounded Model Checker (CBMC) in the same way that mlkem-native is. We plan to formally verify the correctness of the native code using HOL-Light and the [s2n-bignum](https://github.com/awslabs/s2n-bignum) infrastructure - so far this has been completed for the x86_64 NTT (see [#640](https://github.com/pq-code-package/mldsa-native/pull/640)). Both HOL-Light and CBMC proofs run in the mldsa-native CI and should be easy to reproduce.


~**We are about to release the v1.0.0-alpha version, and I will switch to the named release once available**~

**This PR uses the recently released [v1.0.0-alpha](https://github.com/pq-code-package/mldsa-native/releases/tag/v1.0.0-alpha) version of mldsa-native**


- Resolves #2272

mkannwischer commented:
We are nearing the v1.0.0-alpha release and I don't expect any more code changes until then. I'll update to the named release once it's out. Feel free to review already.

mkannwischer commented:
mldsa-native [v1.0.0-alpha](https://github.com/pq-code-package/mldsa-native/releases/tag/v1.0.0-alpha) just landed. I have updated this PR to use that version.

bhess commented on docs/algorithms/sig/ml_dsa.md:
```diff
@@ -24,7 +24,8 @@
 | [Primary Source](#primary-source) | ref                      | All                         | All                             | None                    | True                               | True                                           | False                 |
-| [Primary Source](#primary-source) | avx2                     | x86\_64                     | Darwin,Linux                    | AVX2,POPCNT             | True                               | True                                           | False                 |
+| [Primary Source](#primary-source) | x86\_64                  | x86\_64                     | Darwin,Linux                    | AVX2,BMI2,POPCNT        | True                               | True                                           | False                 |
+| [Primary Source](#primary-source) | aarch64                  | ARM64\_V8                   | Linux,Darwin                    | None                    | True                               | False                                          | False                 |
```
Is this correct, does the arm64 code use no CPU extensions?

bhess reviewed (commented):
Thank you for this work and the integration, @mkannwischer, looks really good!
I really like how smoothly the integration works without needing any patches. The new arm64 optimization and the formally proven memory safety are a big plus. The AVX2 performance penalty seems acceptable to me, given your explanation and the plan to improve it in the next release.

I have two extra questions (besides the inline ones):

- Are HashML-DSA variants available in mldsa-native, or are they planned for the future?
- Could the externalMu variant be used if liboqs decides to support it?

bhess commented:
> Yes, we support both HashML-DSA and externalMu mode. Both are exposed in the public API.
> See https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/mldsa_native.h.
> We match all testvectors that are in the ACVP testsuite - so we have full feature coverage.
> We should be able to add those to liboqs relatively easily.
> Could you take a look at that in a follow-up PR @bhess?

Great to see these features available! As mentioned in yesterday’s OQS call, @xuganyu96 will likely take a closer look at pre-hash in OQS (including for SLH-DSA) as a follow-up.

bhess reviewed (dismissed):
Looks good to me. Thank you very much @mkannwischer for the work on mldsa-native, and the integration to liboqs!

Asking @open-quantum-safe/liboqs-committers for another look.

mkannwischer commented:
> Looks good to me. Thank you very much @mkannwischer for the work on mldsa-native, and the integration to liboqs!
> 
> Asking @open-quantum-safe/liboqs-committers for another look.

Thanks @bhess. Any other comments @dstebila, @baentsch, @alexrow?

baentsch reviewed (dismissed):
> Any other comments @dstebila, @baentsch, @alexrow?

Nope. LGTM.

mkannwischer commented:
I have rebased on top of main (there was a conflict in the README). Can we get this merged?

xuganyu96 commented:
> I have rebased on top of main (there was a conflict in the README). Can we get this merged?

If this pull request is not urgent, I will prioritize pushing this through first thing in 2026. Please let me know if this needs to be merged right away. Happy new year!

xuganyu96 commented:
> Not super urgent. What is currently blocking it?

I am currently away on vacation and don't have the capacity to respond if something goes wrong. I will return on Jan 5.

xuganyu96 commented on docs/algorithms/sig/ml_dsa.yml:
```diff
@@ -37,20 +36,33 @@ parameter-sets:
+    supported-platforms:
+    - architecture: ARM64_V8
+      operating_systems:
+      - Linux
```
The same data field under x86_64 section lists `Darwin` then `Linux`. Here the order is reversed. Having consistent ordering is purely cosmetic but still appreciated.

An alternative is to sort the list in `copy_from_upstream`. However, I am not sure if raising another issue/pull request and spending CI pipeline resources is worth it.

mkannwischer replied:
Cleaned up upstream: https://github.com/pq-code-package/mldsa-native/pull/898

mkannwischer commented:
I have addressed the feedback upstream: https://github.com/pq-code-package/mldsa-native/pull/898.
Apologies for the delay.
Once that is merged, I will create an `alpha2` tag and update this PR. That makes sense anyway, as there have been considerable changes upstream since I created this PR.

bhess reviewed (approved):
LGTM. Ready to merge IMO
